utils/rollout.js

```diff
@@ -0,0 +1,111 @@
+  return allowedRoles.some((role) => {
+    return Boolean(userRoles[`${role}`])
+  })
+}
+
+/**
+ * Rollout the feature to a percentage of users. We use bucket based percentage rollout.
+ * The logic:
+ *   - Take modulus of an identifier with 100
+ *   - Check whether the modulus is greater than the percentage rollout value.
+ * Note:
+ *   - The identifier should be evenly distributed.
+ *   - userId is used as identifier if present in order to make rollout monotonic.
+ *
+ * @param {int} percentValue: Percentage value between 0 to 100
+ * @param {String} userId: Id of the logged in user if present
+ */
+const percentageRollout = (percentValue, userId) => {
+  if (percentValue > 100) {
+    throw Error('Invalid percentage rollout value')
+  }
+
+  const indentifier = userId || generateUUID()
+  const hash = hashCode(indentifier)
+  return (hash % 100) < percentValue
+}
+
+/**
+ * Check the featureFlag and determine whether this flag is enabled for this request.
+ * Use user details if the data is available.
+ *
+ * 3 types of rollout configs are supported (in priority order):
+ *   a. Role based rollout:
+ *     - A list of roles is provided. If the user belongs to any of the roles, then feature is enabled.
+ *       Note: If userData is provided but doesn't match role, we don't fallback to other rollout but disable the feature.
+ *     - If userData is missing, it will fallback to the next type of rollout config.
+ *   b. Percentage rollout:
+ *     - Rollout to a certain percentage of users.
+ *     - User details are used to make this rollout monotonic.
+ *     - If usrData is missing, then the rollout is done randomly.
+ *   c. Toggle rollout:
+ *     - A boolean toggle to either disable or enable the feature.
+ *     - This will the be the value considered when the above rollout are not configured.
+ *
+ * The schema for featureFlag config:
+ * config = {
+ *   roleBased: {
+ *     roles: [Array[String]]<An array of allowed roles - role based rollout>
+ *     active: [Boolean]<whether this rollout method is active>
+ *   },
+ *   percentage: {
+ *     value: [Integer]<percentage of users for which rollout this feature - percentage rollout>
+ *     active: [Boolean]<whether this rollout method is active>
+ *   },
+ *   enabled: [Boolean]<whether the featureFlag is enabled - toggle rollout>
+ * }
+ * @param {Object} featureFlag: Featureflag object from the database.
+ * @param {Object} userData: Details of the user if request is authenticated. Else null.
+ * @return {Boolean}: Whether this feature should be enabled or disabled for this request.
+ */
+const featureFlagRollout = (featureFlag, userData) => {
+  if (!featureFlag.config) {
+    throw Error('Feature flag config is absent')
+  }
+
+  const { roleBased, percentage, enabled } = featureFlag.config
+
+  // Check if roleBased rollout can be used
+  if (roleBased && roleBased.active && userData) {
+    return roleBasedRollout(roleBased.roles || [], userData.roles || {})
```

Comment:
Should the function manage missing values?

Reply:
This is done for defensive reasons.

Sometimes our DB data is not exactly as we expect. Some entries are not migrated and fields might be absent. Hence the required defaults.

Other option is throw exceptions but I think a WARN log should be fine as this is not a critical function.